### PR TITLE
[CP] Optimize Kimi K3 MLA all-gather and Ulysses communication

### PR DESCRIPTION
**Pull request by acisseJZhong**
The original optimization idea was inspired by this [MLA Ulysses CP proposal](https://gist.github.com/fegin/bd9bd9e9147ae1b11375dfb68815f78a).

For both CP backends, this PR keeps the head-shared `k_rope` headless during
communication and expands it across heads locally.

- **Ulysses CP**
  - Before: three input all-to-alls for materialized Q, K, and V. K included
    one repeated `k_rope` copy per head.
  - After: one all-to-all for packed Q and compact per-head K/V, plus one
    all-gather for headless `k_rope`. The output all-to-all is unchanged.
- **K/V all-gather CP**
  - Before: separate all-gathers for materialized K and V. K included repeated
    `k_rope`.
  - After: separate all-gathers for compact per-head K/V and headless `k_rope`,
    followed by local K materialization. Fusing them into one all-gather showed
    no benefit and slightly regressed performance.

## Summary

- Add Kimi K3 MLA Flex and Varlen backends that preserve compact K/V inputs
  through CP communication and materialize K locally.
- Wire model-specific CP backend selection, SPMD layouts, and focused tests.
- CP=8 benchmarks show a `4.03%` K/V all-gather throughput improvement; the
  Ulysses communication reduction does not produce a measurable speedup.

## Tests

Deterministic BF16 runs compared parent commit `5a30c8a1b` with PR commit
`0469340ff` at CP=2, using the same seed checkpoint (`seed=42`), 256 tokens per
step, and 100 training steps. Values come from full-precision TensorBoard
metrics; diffs are relative to the parent implementation.

The optimized collectives change backward reduction ordering, so the training
trajectories are not bitwise identical.

Performance was measured with the Kimi K3 multimodal debug model at CP=8 on
8 H100 GPUs with 2,048 tokens per step. Each run averages steps 21-99 using
full-precision TensorBoard throughput metrics. All-gather results average
three 100-step runs; Ulysses results average five 100-step runs executed in
balanced order.

### All-gather CP: before vs. after

| Step | Generic loss | MLA loss | Diff | Generic grad norm | MLA grad norm | Diff |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | `12.473673` | `12.473673` | `0.000%` | `27.2500` | `27.2500` | `0.000%` |
| 10 | `4.037894` | `4.083763` | `1.136%` | `6.3125` | `5.9062` | `6.436%` |
| 100 | `2.879660` | `2.914417` | `1.207%` | `5.8438` | `5.6562` | `3.209%` |

| Implementation | Mean TPS | Change |
| --- | ---: | ---: |
| Before: `KVAllGatherCPFlexInnerAttention` | `574.3` | |
| After: `KVAllGatherCPMLAFlexInnerAttention` | `597.4` | **`+4.03%`** |

### Ulysses CP: before vs. after

| Step | Generic loss | MLA loss | Diff | Generic grad norm | MLA grad norm | Diff |
| ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| 1 | `12.460487` | `12.460487` | `0.000%` | `27.2500` | `27.3750` | `0.459%` |
| 10 | `4.009133` | `4.295489` | `7.143%` | `7.1875` | `6.2188` | `13.478%` |
| 100 | `2.740203` | `2.837717` | `3.559%` | `5.4688` | `5.4062` | `1.143%` |

| Implementation | Mean TPS +/- run SD | Change |
| --- | ---: | ---: |
| Before: `UlyssesCPFlexInnerAttention` | `607.7 +/- 5.7` | |
| After: `UlyssesCPMLAFlexInnerAttention` | `593.9 +/- 30.9` | `-2.28%` |

### Performance analysis

- **K/V all-gather:** Both implementations use two input all-gathers, while
  the MLA path replaces per-head copies of `k_rope` with one headless tensor.
  Fusing compact K/V and headless `k_rope` into one all-gather showed no
  benefit and slightly regressed throughput, so the implementation keeps two
  all-gathers. Throughput improved by `4.03%`, consistent with the reduced
  communication volume.
- **Ulysses:** Five-run results for each configuration:

  | Configuration | Mean TPS | vs. baseline | Run-to-run SD | Relative variation |
  | --- | ---: | ---: | ---: | ---: |
  | Previous | `607.7` | - | `5.7` | `0.9%` |
  | Packed + headless | `593.9` | `-2.28%` | `30.9` | `5.2%` |
  | Packed + repeated | `609.5` | `+0.29%` | `3.5` | `0.6%` |
  | No-pack + headless | `599.3` | `-1.38%` | `9.5` | `1.6%` |

  Packed Q/KV with repeated `k_rope` matches the baseline, while both headless
  variants are slower on average. The additional all-gather offsets the lower
  `k_rope` payload, and the current packed, headless path also has high
  run-to-run variance. No conclusive Ulysses speedup was measured at this
  scale.

**Note:** This debug-model benchmark is not representative of the full Kimi K3
headless-communication tradeoff. With 16 heads at CP=8, the debug model has
only 2 local heads per rank; the full model has 96 heads, or 12 per rank. The
full model therefore avoids substantially more repeated `k_rope` communication
relative to the extra all-gather.

### CODE REVIEW

File: torchtitan/models/kimi_k3/attention.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+# L = low-rank KV latent dimension (kv_lora_rank)
+# N = per-head non-positional key dimension (qk_nope_head_dim)
+# R = head-shared key dimension (qk_rope_head_dim)
+# V = per-head value dimension (v_head_dim)
+# C = compressed KV channels (L + R)
+# K = full per-head query/key dimension (q_head_dim = N + R)
+# P = packed per-head KV channels (N + V)
+
+
+class MLAInnerAttention(InnerAttention, ABC):
```
Comment (fegin) [inline torchtitan/models/kimi_k3/attention.py:46]: Why putting this for Kimi_k3 only? DSv3 can actually share this. I don't see a difference.

File: torchtitan/models/kimi_k3/attention.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        kv_THP: torch.Tensor,
+        k_shared_TR: torch.Tensor,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """Materialize per-head K/V from MLA's packed KV and shared key."""
+        nope_head_dim = q_THK.shape[-1] - k_shared_TR.shape[-1]
+        value_head_dim = kv_THP.shape[-1] - nope_head_dim
+        k_nope_THN, v_THV = torch.split(kv_THP, [nope_head_dim, value_head_dim], dim=-1)
+        k_shared_THR = k_shared_TR.unsqueeze(1).expand(-1, k_nope_THN.shape[1], -1)
+        k_THK = torch.cat((k_nope_THN, k_shared_THR), dim=-1)
+        return k_THK, v_THV
```
Comment (fegin) [inline torchtitan/models/kimi_k3/attention.py:75]: The only reason why you need this `_materialize_kv` and `MLAInnerAttention` is to support CP. But this actually complicates the code because. We now need to create `MLAFlexInnerAttention` and `MLAVarlenInnerAttention`. 

You can avoid this complication by still putting this logic in the MLAAttention. And do another split and cat in MLACPAttention.  These are duplicated calls, which theoretically should make performance worse but the performance difference should be minimal? The inner attention is will stays as a pure kernel.

Do you think the later approach will make the code cleaner?

File: torchtitan/models/kimi_k3/attention.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        return VarlenInnerAttention.forward(self, q_THK, k_THK, v_THV, **kwargs)
+
+
+class KimiMLAAttention(BaseAttention):
+    """Kimi K3 multi-head latent attention.
+
+    Unlike DeepSeek-V3 MLA, the released K3 configuration sets
+    ``mla_use_nope=True``: the RoPE-sized query/key slices remain part of the
+    projected head, but no rotary transform is applied, so this has no rope
+    config at all. Attention delegates to the configured inner backend.
```
Comment (fegin) [inline torchtitan/models/kimi_k3/attention.py:120]: This doesn't mean that DSv3 cannot share `MLAInnerAttention`

**Review by fegin — CHANGES_REQUESTED** [review]
Put MLAAttention in common if you want to preserve `_materialize_kv` approach.

File: torchtitan/models/kimi_k3/attention.py (hunk 2 above)
Comment (fegin) [reply torchtitan/models/kimi_k3/attention.py:75]: **NOTE:**
I realized you also want to support allgather, which I didn't try before. It seems that using `_materialize_kv` is cleaner with different CP methods.
